url416ee5b4ee_martinmarinov_HonProj_tgz-pJ8-martin_gui_MainJ8/src/martin/quantum/tools/Tools.java

```diff
@@ -0,0 +1,250 @@
+ * which accompanies this distribution, and is available at
+ * http://www.gnu.org/licenses/gpl.html
+ * 
+ * Contributors:
+ *     Martin - initial API and implementation
+ ******************************************************************************/
+package martin.quantum.tools;
+
+import java.io.PrintStream;
+import java.util.ArrayList;
+import java.util.HashMap;
+
+import martin.math.Complex;
+import martin.math.MathsItem;
+
+public final class Tools {
+	
+	/** Default logger to output messages to */
+	public static PrintStream logger = System.out;
+	public static final boolean FORCE_SYSTEM_LOGGER = false;
+	
+	public static boolean SIMPLIFICATION_ENABLED = true;
+	
+	public static boolean VERBOSE = false;
+	public static boolean SILENT = false;
+	public static boolean MATHEMATICA_FRIENDLY_OUTPUT = false;
+	
+	/**
+	 * Parse an expression like "(blah)" to "blah", removes only the outer layer of brackets. Brackets are not compulsory!
+	 * @param input
+	 * @return
+	 */
+	public static String trimAndCheckBrackets(String input) throws Exception {
+		input = input.trim();
+		
+		if (input.isEmpty())
+			return input;
+		
+		if (!balancedBrackets(input))
+			throw new Exception("There is an unbalanced number of brackets in '"+input+"'");
+		
+		if (input.charAt(0) == '(' && input.charAt(input.length()-1) == ')') {
+			final String trimmed = input.substring(1, input.length()-1);
+			
+			if (balancedBrackets(trimmed))
+				return trimmed.trim();
+		}
+		
+		return input;
+	}
+	
+	/**
+	 * Check whether a string contains a balanced set of brackets
+	 * @param input
+	 * @return
+	 */
+	public static boolean balancedBrackets(String input) {
+		int start = 0;
+		final int length = input.length();
+		
+		for (int i = 0; i < length; i++) {
+			char c = input.charAt(i);
+			
+			if (c == ')' || c == '(') start += c == ')' ? -1 : 1;
+			
+			if (start < 0) return false;
+		}
+		return start == 0;
+	}
+	
+	/**
+	 * Splits input by a top level operator. <br/><br/>
+	 * It will transform a string like "(1 + 2*(4 + 5)) + (5 + 6)" into "(1 + 2*(4 + 5)) " and " (5 + 6)"<br/>
+	 * @param input formatted by {@link #trimAndCheckBrackets(String)}
+	 * @param operator the top level operator that we're looking for, in the above case, it's "+". If more than one operator specified, the first one is considered default. Ie "8+2", 8 will have default operator
+	 * @param stackoperators true for + and -. Allows expressions line (1 + 2) + 3 to evaluate to 1 + 2 + 3. Keep in mind for * this should be false since (1 * 3) + 4 * 5 will be undefined in this context
+	 * @return the left and right side around operator or null if top level operator not available. Keep in mind that you would
+	 * need to apply {@link #trimAndCheckBrackets(String)} to the outputs!
+	 * @throws Exception 
+	 */
+	public static ArrayList<Tuple<char[], String>> splitByTopLevel(final String input, char[] operator, boolean stackoperators) throws Exception {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 18. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 90-124, lines 129-159 into separate methods.